File: src/main/java/com/uiptv/model/Channel.java
package com.uiptv.model;

import com.uiptv.shared.BaseJson;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import org.json.JSONObject;

import java.util.Map;
import java.util.stream.Collectors;

import static com.uiptv.util.StringUtils.safeGetString;

@Data
@NoArgsConstructor
@EqualsAndHashCode(callSuper = false)
public class Channel extends BaseJson {
    private static final String KEY_SEASON = "season";

    private String dbId;
    private String channelId;
    private String categoryId;
    private String name;
    private String number;
    private String cmd;
    @SuppressWarnings("java:S116")
    private String cmd_1;
    @SuppressWarnings("java:S116")
    private String cmd_2;
    @SuppressWarnings("java:S116")
    private String cmd_3;
    private String logo;
    private String description;
    private String season;
    private String episodeNum;
    private String releaseDate;
    private String rating;
    private String duration;
    private String extraJson;
    private int censored;
    private int status;
    private int hd;
    private boolean watched;
    private String drmType;
    private String drmLicenseUrl;
    private String clearKeysJson;
    private String inputstreamaddon;
    private String manifestType; // New field for manifest_type

    @SuppressWarnings("java:S107")
    public Channel(String channelId, String name, String number, String cmd, String cmd1, String cmd2, String cmd3, String logo, int censored, int status, int hd, String drmType, String drmLicenseUrl, Map<String, String> clearKeys, String inputstreamaddon, String manifestType) {
        this.channelId = channelId;
        this.name = name;
        this.number = number;
        this.cmd = cmd;
        this.cmd_1 = cmd1;
        this.cmd_2 = cmd2;
        this.cmd_3 = cmd3;
        this.logo = logo;
        this.censored = censored;
        this.status = status;
        this.hd = hd;
        this.drmType = drmType;
        this.drmLicenseUrl = drmLicenseUrl;
        setClearKeys(clearKeys);
        this.inputstreamaddon = inputstreamaddon;
        this.manifestType = manifestType;
    }

    public static Channel fromJson(String json) {
        try {
            JSONObject jsonObj = new JSONObject(json);
            Channel channel = new Channel();
            channel.setDbId(safeGetString(jsonObj, "dbId"));
            channel.setChannelId(safeGetString(jsonObj, "channelId"));
            channel.setCategoryId(safeGetString(jsonObj, "categoryId"));
            channel.setName(safeGetString(jsonObj, "name"));
            channel.setNumber(safeGetString(jsonObj, "number"));
            channel.setCmd(safeGetString(jsonObj, "cmd"));
            channel.setCmd_1(safeGetString(jsonObj, "cmd_1"));
            channel.setCmd_2(safeGetString(jsonObj, "cmd_2"));
            channel.setCmd_3(safeGetString(jsonObj, "cmd_3"));
            channel.setLogo(safeGetString(jsonObj, "logo"));
            channel.setDescription(safeGetString(jsonObj, "description"));
            channel.setSeason(safeGetString(jsonObj, KEY_SEASON));
            channel.setEpisodeNum(safeGetString(jsonObj, "episodeNum"));
            channel.setReleaseDate(safeGetString(jsonObj, "releaseDate"));
            channel.setRating(safeGetString(jsonObj, "rating"));
            channel.setDuration(safeGetString(jsonObj, "duration"));
            channel.setExtraJson(safeGetString(jsonObj, "extraJson"));
            channel.setCensored(jsonObj.optInt("censored"));
            channel.setStatus(jsonObj.optInt("status"));
            channel.setHd(jsonObj.optInt("hd"));
            Object watched = jsonObj.opt("watched");
            if (watched instanceof Boolean watchedBoolean) {
                channel.setWatched(watchedBoolean);
            } else {
                String watchedStr = safeGetString(jsonObj, "watched");
                channel.setWatched("1".equals(watchedStr) || "true".equalsIgnoreCase(watchedStr));
            }
            channel.setDrmType(safeGetString(jsonObj, "drmType"));
            channel.setDrmLicenseUrl(safeGetString(jsonObj, "drmLicenseUrl"));
            channel.setClearKeysJson(safeGetString(jsonObj, "clearKeysJson"));
            channel.setInputstreamaddon(safeGetString(jsonObj, "inputstreamaddon"));
            channel.setManifestType(safeGetString(jsonObj, "manifestType"));
            return channel;
        } catch (Exception _) {
            // Treat malformed serialized channels as absent cached snapshots.
            return null;
        }
    }

    public void setClearKeys(Map<String, String> clearKeys) {
        if (clearKeys == null || clearKeys.isEmpty()) {
            this.clearKeysJson = null;
            return;
        }
        this.clearKeysJson = clearKeys.entrySet()
                .stream()
                .map(entry -> "\"" + entry.getKey() + "\":\"" + entry.getValue() + "\"")
                .collect(Collectors.joining(",", "{", "}"));
    }

    public int getCompareSeason() {
        try {
            return Integer.parseInt(this.getName().split("-")[0]
                    .toLowerCase()
                    .replace(" ", "")
                    .replace(KEY_SEASON, "")
                    .replace("episode", "")
                    .replace("-", ""));
        } catch (Exception _) {
            // Fall back to zero when the title cannot be parsed into a season number.
        }
        return 0;
    }

    public int getCompareEpisode() {
        try {
            return Integer.parseInt(this.getName().split("-")[1]
                    .toLowerCase()
                    .replace(" ", "")
                    .replace(KEY_SEASON, "")
                    .replace("episode", "")
                    .replace("-", ""));
        } catch (Exception _) {
            // Fall back to zero when the title cannot be parsed into an episode number.
        }
        return 0;
    }
}

File: server/src/main/kotlin/com/uiptv/model/Channel.kt
package com.uiptv.model

import com.uiptv.shared.BaseJson
import java.util.stream.Collectors

import com.uiptv.util.StringUtils.safeGetString
import com.uiptv.util.json.optBoolean
import com.uiptv.util.json.optInt
import com.uiptv.util.json.parseJsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull

private const val KEY_SEASON = "season"

data class Channel @JvmOverloads constructor(
    var dbId: String? = null,
    var channelId: String? = null,
    var categoryId: String? = null,
    var name: String? = null,
    var number: String? = null,
    var cmd: String? = null,
    @Suppress("PropertyName")
    var cmd_1: String? = null,
    @Suppress("PropertyName")
    var cmd_2: String? = null,
    @Suppress("PropertyName")
    var cmd_3: String? = null,
    var logo: String? = null,
    var description: String? = null,
    var season: String? = null,
    var episodeNum: String? = null,
    var releaseDate: String? = null,
    var rating: String? = null,
    var duration: String? = null,
    var extraJson: String? = null,
    var censored: Int = 0,
    var status: Int = 0,
    var hd: Int = 0,
    @get:JvmName("isWatched")
    var watched: Boolean = false,
    var drmType: String? = null,
    var drmLicenseUrl: String? = null,
    var clearKeysJson: String? = null,
    var inputstreamaddon: String? = null,
    var manifestType: String? = null
) : BaseJson() {
    @Suppress("java:S107")
    constructor(
        channelId: String?,
        name: String?,
        number: String?,
        cmd: String?,
        cmd1: String?,
        cmd2: String?,
        cmd3: String?,
        logo: String?,
        censored: Int,
        status: Int,
        hd: Int,
        drmType: String?,
        drmLicenseUrl: String?,
        clearKeys: Map<String, String>?,
        inputstreamaddon: String?,
        manifestType: String?
    ) : this(
        channelId = channelId,
        name = name,
        number = number,
        cmd = cmd,
        cmd_1 = cmd1,
        cmd_2 = cmd2,
        cmd_3 = cmd3,
        logo = logo,
        censored = censored,
        status = status,
        hd = hd,
        drmType = drmType,
        drmLicenseUrl = drmLicenseUrl,
        inputstreamaddon = inputstreamaddon,
        manifestType = manifestType
    ) {
        setClearKeys(clearKeys)
    }

    fun setClearKeys(clearKeys: Map<String, String>?) {
        clearKeysJson = if (clearKeys.isNullOrEmpty()) {
            null
        } else {
            clearKeys.entries.stream()
                .map { entry -> "\"${entry.key}\":\"${entry.value}\"" }
                .collect(Collectors.joining(",", "{", "}"))
        }
    }

    fun getCompareSeason(): Int {
        return try {
            name!!.split("-")[0]
                .lowercase()
                .replace(" ", "")
                .replace(KEY_SEASON, "")
                .replace("episode", "")
                .replace("-", "")
                .toInt()
        } catch (_: Exception) {
            0
        }
    }

    fun getCompareEpisode(): Int {
        return try {
            name!!.split("-")[1]
                .lowercase()
                .replace(" ", "")
                .replace(KEY_SEASON, "")
                .replace("episode", "")
                .replace("-", "")
                .toInt()
        } catch (_: Exception) {
            0
        }
    }

    companion object {
        @JvmStatic
        fun fromJson(json: String): Channel? {
            return try {
                val jsonObj = parseJsonObject(json) ?: return null
                val channel = Channel(
                    dbId = safeGetString(jsonObj, "dbId"),
                    channelId = safeGetString(jsonObj, "channelId"),
                    categoryId = safeGetString(jsonObj, "categoryId"),
                    name = safeGetString(jsonObj, "name"),
                    number = safeGetString(jsonObj, "number"),
                    cmd = safeGetString(jsonObj, "cmd"),
                    cmd_1 = safeGetString(jsonObj, "cmd_1"),
                    cmd_2 = safeGetString(jsonObj, "cmd_2"),
                    cmd_3 = safeGetString(jsonObj, "cmd_3"),
                    logo = safeGetString(jsonObj, "logo"),
                    description = safeGetString(jsonObj, "description"),
                    season = safeGetString(jsonObj, KEY_SEASON),
                    episodeNum = safeGetString(jsonObj, "episodeNum"),
                    releaseDate = safeGetString(jsonObj, "releaseDate"),
                    rating = safeGetString(jsonObj, "rating"),
                    duration = safeGetString(jsonObj, "duration"),
                    extraJson = safeGetString(jsonObj, "extraJson"),
                    censored = jsonObj.optInt("censored"),
                    status = jsonObj.optInt("status"),
                    hd = jsonObj.optInt("hd"),
                    drmType = safeGetString(jsonObj, "drmType"),
                    drmLicenseUrl = safeGetString(jsonObj, "drmLicenseUrl"),
                    clearKeysJson = safeGetString(jsonObj, "clearKeysJson"),
                    inputstreamaddon = safeGetString(jsonObj, "inputstreamaddon"),
                    manifestType = safeGetString(jsonObj, "manifestType")
                )
                val watched = jsonObj["watched"]
                channel.watched = when (watched) {
                    is JsonPrimitive -> watched.booleanOrNull ?: safeGetString(jsonObj, "watched").let {
                        "1".equals(it) || "true".equals(it, ignoreCase = true)
                    }
                    else -> {
                        val watchedStr = safeGetString(jsonObj, "watched")
                        "1".equals(watchedStr) || "true".equals(watchedStr, ignoreCase = true)
                    }
                }
                channel
            } catch (_: Exception) {
                null
            }
        }
    }
}
